client/client-multi/client-c/src/kaa/platform-impl/posix/key_utils.c

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright 2014-2016 CyberVision, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/stat.h>
+#include <string.h>
+#include <utilities/kaa_mem.h>
+#include <utilities/kaa_aes_rsa.h>
+#include <platform/ext_key_utils.h>
+#include <mbedtls/pk.h>
+#include <mbedtls/entropy.h>
+#include <mbedtls/ctr_drbg.h>
+#include <mbedtls/md.h>
+
+#include <kaa_rsa_key_gen.h>
+
+#define ENC_SESSION_KEY_LENGTH         256
+#define KAA_SIGNATURE_LENGTH           256
+#define AES_ECB_ENCRYPTION_CHUNCK_SIZE 16
+#define KAA_RSA_PUBLIC_KEY_LENGTH_MAX  294
+#define KAA_SHA1_PUB_LEN               20
+#define KAA_SHA1_PUB_BASE64_LEN        28
+
+#define KAA_PRIVATE_KEY_STORAGE "private.key"
+#define KAA_PUBLIC_KEY_STORAGE "public.key"
+
+/**
+ * Contains Endpoint keys.
+ *
+ * The main purpose of the structure is
+ * caching calculated keys.
+ */
+
+typedef struct {
+    uint8_t session_key[KAA_SESSION_KEY_LENGTH];
+    uint8_t encrypted_session_key[ENC_SESSION_KEY_LENGTH];
+    uint8_t signature[KAA_SIGNATURE_LENGTH];
+    size_t  encrypted_session_key_length;
+    size_t  session_key_length;
+} endpoint_key_utils_t;
+
+/* Endpoint's RSA Keys */
+static endpoint_key_utils_t keys;
+static mbedtls_pk_context pk_context_;
+static mbedtls_pk_context pk_pub_context;
+
+/* TODO(KAA-1089) */
+__attribute__((weak))
+void ext_get_endpoint_public_key(uint8_t **buffer, size_t *buffer_size)
+{
+    if (!buffer || !buffer_size) {
+        return;
+    }
+    static int key_length;
```

Comment:
key_length can be negative after:
key_length = mbedtls_pk_write_pubkey_der(&pk_pub_context, buff, sizeof(buff));